orchagent/pfcactionhandler.cpp

```diff
@@ -0,0 +1,466 @@
+#include "pfcactionhandler.h"
+#include "logger.h"
+#include "saiserialize.h"
+
+#include <vector>
+
+#define PFC_WD_QUEUE_STATUS             "PFC_WD_STATUS"
+#define PFC_WD_QUEUE_STATUS_OPERATIONAL "operational"
+#define PFC_WD_QUEUE_STATUS_STORMED     "stormed"
+
+#define PFC_WD_QUEUE_STATS_DEADLOCK_DETECTED "PFC_WD_QUEUE_STATS_DEADLOCK_DETECTED"
+#define PFC_WD_QUEUE_STATS_DEADLOCK_RESTORED "PFC_WD_QUEUE_STATS_DEADLOCK_RESTORED"
+
+#define SAI_QUEUE_STAT_PACKETS_STR         "SAI_QUEUE_STAT_PACKETS"
+#define SAI_QUEUE_STAT_DROPPED_PACKETS_STR "SAI_QUEUE_STAT_DROPPED_PACKETS"
+
+extern sai_object_id_t gSwitchId;
+extern sai_port_api_t *sai_port_api;
+extern sai_queue_api_t *sai_queue_api;
+extern sai_buffer_api_t *sai_buffer_api;
+
+PfcWdActionHandler::PfcWdActionHandler(sai_object_id_t port, sai_object_id_t queue,
+        uint8_t queueId, shared_ptr<Table> countersTable):
+    m_port(port),
+    m_queue(queue),
+    m_queueId(queueId),
+    m_countersTable(countersTable)
+{
+    SWSS_LOG_ENTER();
+
+    SWSS_LOG_NOTICE(
+            "PFC Watchdog detected PFC storm on queue 0x%lx port 0x%lx",
+            m_queue,
+            m_port);
+    
+    m_stats = getQueueStats(m_countersTable, sai_serialize_object_id(m_queue));
+    m_stats.detectCount++;
+    m_stats.operational = false;
+
+    updateWdCounters(sai_serialize_object_id(m_queue), m_stats);
+}
+
+PfcWdActionHandler::~PfcWdActionHandler(void)
+{
+    SWSS_LOG_ENTER();
+
+    auto finalStats = getQueueStats(m_countersTable, sai_serialize_object_id(m_queue));
+
+    SWSS_LOG_NOTICE(
+            "Queue 0x%lx port 0x%lx restored from PFC storm. Tx packets: %lu. Dropped packets: %lu",
+            m_queue,
+            m_port,
+            finalStats.txPkt - m_stats.txPkt,
+            finalStats.txDropPkt - m_stats.txDropPkt);
+
+    finalStats.restoreCount++;
```

Comment:
Are we also going to log drop pkts in database? We need to log accumulative ingress and egress drops and ingress and egress drops since last shutdown action.

Reply:
Will be in phase 2

Reply:
ok